File: Bot/CompleteTypeScript/src/GameConnection.ts
import { ChannelCredentials } from "@grpc/grpc-js";
import { GrpcTransport } from "@protobuf-ts/grpc-transport";
import { Game } from "./Game";
import { GameResultError } from "./GameResultError";
import { Result } from "./generated/swoq";
import { GameServiceClient, IGameServiceClient } from "./generated/swoq.client";

export class GameConnection {
    private readonly client: IGameServiceClient;

    public readonly userId: string;

    constructor(host: string, userId: string) {
        this.userId = userId;
        const transport = new GrpcTransport({
            host,
            channelCredentials: ChannelCredentials.createInsecure(),
        });
        this.client = new GameServiceClient(transport);
    }

    public async start(level?: number): Promise<Game> {
        const { response } = await this.client.start({
            userId: this.userId,
            level,
        });
        // TODO Add queued quest handling
        if (response.result !== Result.OK) {
            throw new GameResultError(response.result);
        }
        if (
            response.gameId === undefined ||
            response.width === undefined ||
            response.height === undefined ||
            response.visibilityRange === undefined ||
            response.state === undefined
        ) {
            throw new Error("assertion failed: invalid start response");
        }
        return new Game(
            this.client,
            response.gameId,
            response.width,
            response.height,
            response.visibilityRange,
            response.state
        );
    }
}

File: Bot/ExampleGo/game_connection.go
package main

import (
	"bot/proto/swoq"
	"context"
	"fmt"
	"log"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

type GameConnection struct {
	userId        string
	userName      string
	replaysFolder string
	conn          *grpc.ClientConn
	client        swoq.GameServiceClient
}

func NewGameConnection(host string, userId string, userName string, replaysFolder string) (*GameConnection, error) {
	conn, err := grpc.NewClient(host, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, err
	}
	client := swoq.NewGameServiceClient(conn)

	return &GameConnection{userId: userId, userName: userName, replaysFolder: replaysFolder, conn: conn, client: client}, nil
}

func (g GameConnection) Close() error {
	return g.conn.Close()
}

func (g GameConnection) Start(level *int32, seed *int32) (*Game, error) {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()
	startRequest := &swoq.StartRequest{UserId: g.userId, UserName: g.userName, Level: level, Seed: seed}

	startResponse, err := g.client.Start(ctx, startRequest)
	if err != nil {
		return nil, err
	}
	for startResponse.Result == swoq.StartResult_START_RESULT_QUEST_QUEUED {
		log.Printf("Quest queued, retrying ...")
		startResponse, err = g.client.Start(ctx, startRequest)
		if err != nil {
			return nil, err
		}
	}
	if startResponse.Result != swoq.StartResult_START_RESULT_OK {
		return nil, fmt.Errorf("Start failed (result %s)", startResponse.Result.String())
	}

	var replayFile *ReplayFile = nil
	if g.replaysFolder != "" {
		var err error
		replayFile, err = NewReplayFile(g.replaysFolder, startRequest, startResponse)
		if err != nil {
			return nil, err
		}
	}

	return &Game{
		client:          &g.client,
		gameId:          *startResponse.GameId,
		mapWidth:        *startResponse.MapWidth,
		mapHeight:       *startResponse.MapHeight,
		visibilityRange: *startResponse.VisibilityRange,
		state:           startResponse.State,
		seed:            startResponse.Seed,
		replayFile:      replayFile,
	}, nil
}
